{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee4a21c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Count': '1', 'RetMax': '1', 'RetStart': '0', 'IdList': ['168994'], 'TranslationSet': [], 'TranslationStack': [{'Term': 'PRJNA168994[All Fields]', 'Field': 'All Fields', 'Count': '1', 'Explode': 'N'}, 'GROUP'], 'QueryTranslation': 'PRJNA168994[All Fields]'}\n",
      "UID del BioProject: 168994\n"
     ]
    }
   ],
   "source": [
    "# Cargando el módulo Entrez de la biblioteca biopython y otras bibliotecas necesarias para esta sesión\n",
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import subprocess\n",
    "import os\n",
    "\n",
    "# Proporcionar correo electrónico\n",
    "Entrez.email = \"yeimicc@lcg.unam.mx\"\n",
    "# Buscar el UID del BioProject\n",
    "project_acc = \"PRJNA168994\"\n",
    "handle = Entrez.esearch(db=\"bioproject\", term=project_acc)\n",
    "search_results = Entrez.read(handle)\n",
    "print(search_results)\n",
    "handle.close()\n",
    "# Extraer solo el UID\n",
    "project_uid = search_results[\"IdList\"][0]\n",
    "print(\"UID del BioProject:\", project_uid)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "389794c3",
   "metadata": {},
   "source": [
    "* Bio.Entrez es un modulo de biopiton para interactuar con las cases de datos de NCBI\n",
    "* pandas para manipular datos tabulares\n",
    "* StringIO para tratar cadenas como archivos\n",
    "* subprocess para ejecutar comandos del sistema desde Python\n",
    "* os para interactuar con el sistema operativo (rutas o archivos)\n",
    "\n",
    "NCBI requiere un correo electronico para rastrear el uso de su API, obligatorio para usar Entrez\n",
    "\n",
    "project_acc: es el num de acceso del BioProject a consultar\n",
    "Entrez.esearch busca en la base de datos bioproject usando el ID\n",
    "handle es un archivo temporal, formato XML, tienes que leer el contenido y convertirlo a un diccionario de Python\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "34014e17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'TaxId': '7227', 'Project_Id': '168994', 'Project_Acc': 'PRJNA168994', 'Project_Type': 'Primary submission', 'Project_Data_Type': 'Transcriptome or Gene expression', 'Sort_By_ProjectType': '889987', 'Sort_By_DataType': '887347', 'Sort_By_Organism': '130844', 'Project_Subtype': '', 'Project_Target_Scope': 'Multiisolate', 'Project_Target_Material': 'Transcriptome', 'Project_Target_Capture': 'Whole', 'Project_MethodType': 'Array', 'Project_Method': '', 'Project_Objectives_List': [{'Project_ObjectivesType': 'Expression', 'Project_Objectives': ''}], 'Registration_Date': '2012/06/20 00:00', 'Project_Name': 'Drosophila melanogaster', 'Project_Title': 'Drosophila melanogaster Transcriptome or Gene expression', 'Project_Description': 'RNA profiling data sets generated by the Drosophila modENCODE project.', 'Keyword': '', 'Relevance_Agricultural': '', 'Relevance_Medical': '', 'Relevance_Industrial': '', 'Relevance_Environmental': '', 'Relevance_Evolution': '', 'Relevance_Model': 'yes', 'Relevance_Other': '', 'Organism_Name': 'Drosophila melanogaster', 'Organism_Strain': '', 'Organism_Label': '', 'Sequencing_Status': 'SRA/Trace', 'Submitter_Organization': 'modENCODE DCC', 'Submitter_Organization_List': ['modENCODE DCC'], 'Supergroup': 'Eukaryotes'}, attributes={'uid': '168994'})], 'DbBuild': 'Build250930-0720.1'}, attributes={'status': 'OK'})}\n"
     ]
    }
   ],
   "source": [
    "# Obtener resumen del BioProject\n",
    "handle = Entrez.esummary(db=\"bioproject\", id=project_uid)\n",
    "proj_record = Entrez.read(handle)\n",
    "print(proj_record)\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c1b77488",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['TaxId', 'Project_Id', 'Project_Acc', 'Project_Type', 'Project_Data_Type', 'Sort_By_ProjectType', 'Sort_By_DataType', 'Sort_By_Organism', 'Project_Subtype', 'Project_Target_Scope', 'Project_Target_Material', 'Project_Target_Capture', 'Project_MethodType', 'Project_Method', 'Project_Objectives_List', 'Registration_Date', 'Project_Name', 'Project_Title', 'Project_Description', 'Keyword', 'Relevance_Agricultural', 'Relevance_Medical', 'Relevance_Industrial', 'Relevance_Environmental', 'Relevance_Evolution', 'Relevance_Model', 'Relevance_Other', 'Organism_Name', 'Organism_Strain', 'Organism_Label', 'Sequencing_Status', 'Submitter_Organization', 'Submitter_Organization_List', 'Supergroup'])\n",
      "\n",
      "=== Información del BioProject ===\n",
      "Acceso: PRJNA168994\n",
      "Título: Drosophila melanogaster Transcriptome or Gene expression\n",
      "Organismo: Drosophila melanogaster\n",
      "Descripción: RNA profiling data sets generated by the Drosophila modENCODE project.\n"
     ]
    }
   ],
   "source": [
    "# Seleccionando la información relacionada al proyecto\n",
    "proj = proj_record[\"DocumentSummarySet\"][\"DocumentSummary\"][0]\n",
    "\n",
    "# Mostrando todos los campos del diccionario\n",
    "print(proj.keys())\n",
    "\n",
    "# Accediendo a algunos campos\n",
    "print(\"\\n=== Información del BioProject ===\")\n",
    "print(\"Acceso:\", proj[\"Project_Acc\"])\n",
    "print(\"Título:\", proj[\"Project_Title\"])\n",
    "print(\"Organismo:\", proj[\"Organism_Name\"])\n",
    "print(\"Descripción:\", proj.get(\"Project_Description\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e89f7bde",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "El proyecto PRJNA168994  está asociado a la base de datos de GEO\n",
      "UIDs GEO asociados: ['200040045', '200040043', '200040042', '200040040', '200040039'] ... total: 156\n"
     ]
    }
   ],
   "source": [
    "# Revisar enlaces desde BioProject hacia GEO (GDS), con el uid del bioproject\n",
    "handle = Entrez.elink(dbfrom=\"bioproject\", db=\"gds\", id=project_uid)\n",
    "linksBioProj = Entrez.read(handle)\n",
    "handle.close()\n",
    "\n",
    "#print(linksBioProj)\n",
    "\n",
    "if linksBioProj[0][\"LinkSetDb\"]:\n",
    "    geo_ids = [link[\"Id\"] for link in linksBioProj[0][\"LinkSetDb\"][0][\"Link\"]]\n",
    "    print(\"\\nEl proyecto\", project_acc, \" está asociado a la base de datos de GEO\")\n",
    "    print(\"UIDs GEO asociados:\", geo_ids[:5], \"... total:\", len(geo_ids))\n",
    "else:\n",
    "    print(\"\\nEl proyecto\", project_acc, \" NO tiene enlaces directos a GEO\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2a258e6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Item': [], 'Id': '200040043', 'Accession': 'GSE40043', 'GDS': '', 'title': 'Drosophila melanogaster Pupae 2-4 days', 'summary': 'modENCODE_submission_757 This submission comes from a modENCODE project of Eric Lai. For full list of modENCODE projects, see http://www.genome.gov/26524648 Project Goal: We plan to generate a comprehensive catalog of expressed and functional microRNAs, and generate biological evidence for their regulatory activity. We plan also to delineate the primary transcription units of microRNA genes. Finally, we plan to annotate other classes of non-miRNA expressed small RNAs, as least some of which may define novel classes of small RNA genes. For data usage terms and conditions, please refer to http://www.genome.gov/27528022 and http://www.genome.gov/Pages/Research/ENCODE/ENCODEDataReleasePolicyFinal2008.pdf', 'GPL': '9058', 'GSE': '40043', 'taxon': 'Drosophila melanogaster', 'entryType': 'GSE', 'gdsType': 'Non-coding RNA profiling by high throughput sequencing', 'ptechType': '', 'valType': '', 'SSInfo': '', 'subsetInfo': '', 'PDAT': '2012/08/10', 'suppFile': '', 'Samples': [{'Accession': 'GSM984194', 'Title': 'Pupae 2-4d (Replicate 1) extraction1_seq1 aliquote 1'}, {'Accession': 'GSM984195', 'Title': 'Pupae 2-4d (Replicate 2) extraction2_seq1 aliquote 1'}], 'Relations': [], 'ExtRelations': [{'RelationType': 'SRA', 'TargetObject': 'SRX002668', 'TargetFTPLink': 'ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRX/SRX002/SRX002668/'}, {'RelationType': 'SRA', 'TargetObject': 'SRX002670', 'TargetFTPLink': 'ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRX/SRX002/SRX002670/'}], 'n_samples': IntegerElement(2, attributes={}), 'SeriesTitle': '', 'PlatformTitle': '', 'PlatformTaxa': '', 'SamplesTaxa': '', 'PubMedIds': [], 'Projects': [{'Name': 'modENCODE', 'URL': 'https://www.ncbi.nlm.nih.gov/geo/encode/'}], 'FTPLink': 'ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE40nnn/GSE40043/', 'GEO2R': 'no'}\n",
      "dict_keys(['Item', 'Id', 'Accession', 'GDS', 'title', 'summary', 'GPL', 'GSE', 'taxon', 'entryType', 'gdsType', 'ptechType', 'valType', 'SSInfo', 'subsetInfo', 'PDAT', 'suppFile', 'Samples', 'Relations', 'ExtRelations', 'n_samples', 'SeriesTitle', 'PlatformTitle', 'PlatformTaxa', 'SamplesTaxa', 'PubMedIds', 'Projects', 'FTPLink', 'GEO2R'])\n"
     ]
    }
   ],
   "source": [
    "# Verificando la información de GEO\n",
    "geo_uid = geo_ids[1]  # ejemplo, segundo UID\n",
    "handle = Entrez.esummary(db=\"gds\", id=geo_uid)\n",
    "summary = Entrez.read(handle)\n",
    "handle.close()\n",
    "print(summary[0])\n",
    "print(summary[0].keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "368af028",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== GEO Accession ===\n",
      "Accession: GSE40043\n",
      "Tipo: GSE\n",
      "Título: Drosophila melanogaster Pupae 2-4 days\n",
      "Resumen: modENCODE_submission_757 This submission comes from a modENCODE project of Eric Lai. For full list of modENCODE projects, see http://www.genome.gov/26524648 Project Goal: We plan to generate a comprehensive catalog of expressed and functional microRNAs, and generate biological evidence for their regulatory activity. We plan also to delineate the primary transcription units of microRNA genes. Finally, we plan to annotate other classes of non-miRNA expressed small RNAs, as least some of which may define novel classes of small RNA genes. For data usage terms and conditions, please refer to http://www.genome.gov/27528022 and http://www.genome.gov/Pages/Research/ENCODE/ENCODEDataReleasePolicyFinal2008.pdf\n",
      "Organismo: Drosophila melanogaster\n",
      "Número de muestras: IntegerElement(2, attributes={})\n",
      "PubMed IDs: []\n",
      "Enlace FTP: ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE40nnn/GSE40043/\n",
      "Archivos suplementarios: \n",
      "Proyectos:  [{'Name': 'modENCODE', 'URL': 'https://www.ncbi.nlm.nih.gov/geo/encode/'}]\n",
      "\n",
      "=== Muestras ===\n",
      "GSM984194 : Pupae 2-4d (Replicate 1) extraction1_seq1 aliquote 1\n",
      "GSM984195 : Pupae 2-4d (Replicate 2) extraction2_seq1 aliquote 1\n"
     ]
    }
   ],
   "source": [
    "# imprimiendo los campos\n",
    "geo_record=summary[0]\n",
    "print(\"=== GEO Accession ===\")\n",
    "print(\"Accession:\", geo_record.get(\"Accession\"))\n",
    "print(\"Tipo:\", geo_record.get(\"entryType\"))\n",
    "print(\"Título:\", geo_record.get(\"title\"))\n",
    "print(\"Resumen:\", geo_record.get(\"summary\"))\n",
    "print(\"Organismo:\", geo_record.get(\"taxon\"))\n",
    "print(\"Número de muestras:\", geo_record.get(\"n_samples\"))\n",
    "print(\"PubMed IDs:\", geo_record.get(\"PubMedIds\"))\n",
    "print(\"Enlace FTP:\", geo_record.get(\"FTPLink\"))\n",
    "print(\"Archivos suplementarios:\", geo_record.get(\"suppFile\"))\n",
    "print(\"Proyectos: \",geo_record.get(\"Projects\"))\n",
    "\n",
    "# Muestras asociadas\n",
    "print(\"\\n=== Muestras ===\")\n",
    "for sample in geo_record.get(\"Samples\",[]):\n",
    "    print(sample['Accession'],\":\",sample['Title'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f81b4706",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Count': '1', 'RetMax': '1', 'RetStart': '0', 'IdList': ['14790'], 'TranslationSet': [], 'TranslationStack': [{'Term': 'GSM461177[All Fields]', 'Field': 'All Fields', 'Count': '1', 'Explode': 'N'}, 'GROUP'], 'QueryTranslation': 'GSM461177[All Fields]'}\n"
     ]
    }
   ],
   "source": [
    "# Obtener el UID correspondiente a la muestra\n",
    "gsm_id = \"GSM461177\"\n",
    "handle = Entrez.esearch(db=\"sra\", term=gsm_id)\n",
    "record = Entrez.read(handle)\n",
    "print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "883772f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Run', 'ReleaseDate', 'LoadDate', 'spots', 'bases', 'spots_with_mates',\n",
      "       'avgLength', 'size_MB', 'AssemblyName', 'download_path', 'Experiment',\n",
      "       'LibraryName', 'LibraryStrategy', 'LibrarySelection', 'LibrarySource',\n",
      "       'LibraryLayout', 'InsertSize', 'InsertDev', 'Platform', 'Model',\n",
      "       'SRAStudy', 'BioProject', 'Study_Pubmed_id', 'ProjectID', 'Sample',\n",
      "       'BioSample', 'SampleType', 'TaxID', 'ScientificName', 'SampleName',\n",
      "       'g1k_pop_code', 'source', 'g1k_analysis_group', 'Subject_ID', 'Sex',\n",
      "       'Disease', 'Tumor', 'Affection_Status', 'Analyte_Type',\n",
      "       'Histological_Type', 'Body_Site', 'CenterName', 'Submission',\n",
      "       'dbgap_study_accession', 'Consent', 'RunHash', 'ReadHash'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Platform</th>\n",
       "      <th>LibraryName</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>Sample</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>SampleName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR031714</td>\n",
       "      <td>SRX014459</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>S2_DRSC_Untreated-3</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>SRS008447</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>GSM461177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR031715</td>\n",
       "      <td>SRX014459</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>S2_DRSC_Untreated-3</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>SRS008447</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>GSM461177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Run Experiment  Platform          LibraryName LibraryLayout  \\\n",
       "0  SRR031714  SRX014459  ILLUMINA  S2_DRSC_Untreated-3        PAIRED   \n",
       "1  SRR031715  SRX014459  ILLUMINA  S2_DRSC_Untreated-3        PAIRED   \n",
       "\n",
       "      Sample           ScientificName SampleName  \n",
       "0  SRS008447  Drosophila melanogaster  GSM461177  \n",
       "1  SRS008447  Drosophila melanogaster  GSM461177  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Consultar si existe el identificador\n",
    "if not record[\"IdList\"]:\n",
    "    print(\"No se encontraron SRR para\", gsm_id)\n",
    "else:\n",
    "    sra_id = record[\"IdList\"]\n",
    "# Obtener la información de la corrida como texto\n",
    "    handle = Entrez.efetch(db=\"sra\", id=sra_id, rettype=\"runinfo\", retmode=\"text\")\n",
    "    runinfo = handle.read()\n",
    "    handle.close()\n",
    "    \n",
    "    if isinstance(runinfo, bytes):\n",
    "        runinfo = runinfo.decode(\"utf-8\")\n",
    "\n",
    "# Cargar en pandas el resultado, directamente desde la cadena\n",
    "    df = pd.read_csv(StringIO(runinfo))\n",
    "    print(df.columns)\n",
    "    display(df[['Run','Experiment','Platform','LibraryName','LibraryLayout','Sample','ScientificName','SampleName']].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1ff5f7e",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[WinError 2] El sistema no puede encontrar el archivo especificado",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[12]\u001b[39m\u001b[32m, line 5\u001b[39m\n\u001b[32m      2\u001b[39m output_dir = \u001b[33m\"\u001b[39m\u001b[33m/home/lvega/data\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      4\u001b[39m \u001b[38;5;66;03m# Hacer el prefetch de los archivos srr\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m5\u001b[39m \u001b[43msubprocess\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mprefetch\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m--output-directory\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moutput_dir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mSRR031714\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[32m      6\u001b[39m subprocess.run([\u001b[33m\"\u001b[39m\u001b[33mprefetch\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33m--output-directory\u001b[39m\u001b[33m\"\u001b[39m, output_dir, \u001b[33m\"\u001b[39m\u001b[33mSRR031715\u001b[39m\u001b[33m\"\u001b[39m], check=\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mC:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.13_3.13.2032.0_x64__qbz5n2kfra8p0\\Lib\\subprocess.py:554\u001b[39m, in \u001b[36mrun\u001b[39m\u001b[34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[39m\n\u001b[32m    551\u001b[39m     kwargs[\u001b[33m'\u001b[39m\u001b[33mstdout\u001b[39m\u001b[33m'\u001b[39m] = PIPE\n\u001b[32m    552\u001b[39m     kwargs[\u001b[33m'\u001b[39m\u001b[33mstderr\u001b[39m\u001b[33m'\u001b[39m] = PIPE\n\u001b[32m--> \u001b[39m\u001b[32m554\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mPopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43mpopenargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m process:\n\u001b[32m    555\u001b[39m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m    556\u001b[39m         stdout, stderr = process.communicate(\u001b[38;5;28minput\u001b[39m, timeout=timeout)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mC:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.13_3.13.2032.0_x64__qbz5n2kfra8p0\\Lib\\subprocess.py:1039\u001b[39m, in \u001b[36mPopen.__init__\u001b[39m\u001b[34m(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, user, group, extra_groups, encoding, errors, text, umask, pipesize, process_group)\u001b[39m\n\u001b[32m   1035\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.text_mode:\n\u001b[32m   1036\u001b[39m             \u001b[38;5;28mself\u001b[39m.stderr = io.TextIOWrapper(\u001b[38;5;28mself\u001b[39m.stderr,\n\u001b[32m   1037\u001b[39m                     encoding=encoding, errors=errors)\n\u001b[32m-> \u001b[39m\u001b[32m1039\u001b[39m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_execute_child\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexecutable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpreexec_fn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mclose_fds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1040\u001b[39m \u001b[43m                        \u001b[49m\u001b[43mpass_fds\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcwd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43menv\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1041\u001b[39m \u001b[43m                        \u001b[49m\u001b[43mstartupinfo\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcreationflags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mshell\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1042\u001b[39m \u001b[43m                        \u001b[49m\u001b[43mp2cread\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mp2cwrite\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1043\u001b[39m \u001b[43m                        \u001b[49m\u001b[43mc2pread\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mc2pwrite\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1044\u001b[39m \u001b[43m                        \u001b[49m\u001b[43merrread\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merrwrite\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1045\u001b[39m \u001b[43m                        \u001b[49m\u001b[43mrestore_signals\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1046\u001b[39m \u001b[43m                        \u001b[49m\u001b[43mgid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgids\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43muid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mumask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1047\u001b[39m \u001b[43m                        \u001b[49m\u001b[43mstart_new_session\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprocess_group\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1048\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[32m   1049\u001b[39m     \u001b[38;5;66;03m# Cleanup if the child failed starting.\u001b[39;00m\n\u001b[32m   1050\u001b[39m     \u001b[38;5;28;01mfor\u001b[39;00m f \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mfilter\u001b[39m(\u001b[38;5;28;01mNone\u001b[39;00m, (\u001b[38;5;28mself\u001b[39m.stdin, \u001b[38;5;28mself\u001b[39m.stdout, \u001b[38;5;28mself\u001b[39m.stderr)):\n",
      "\u001b[36mFile \u001b[39m\u001b[32mC:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.13_3.13.2032.0_x64__qbz5n2kfra8p0\\Lib\\subprocess.py:1554\u001b[39m, in \u001b[36mPopen._execute_child\u001b[39m\u001b[34m(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, unused_restore_signals, unused_gid, unused_gids, unused_uid, unused_umask, unused_start_new_session, unused_process_group)\u001b[39m\n\u001b[32m   1552\u001b[39m \u001b[38;5;66;03m# Start the process\u001b[39;00m\n\u001b[32m   1553\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m1554\u001b[39m     hp, ht, pid, tid = \u001b[43m_winapi\u001b[49m\u001b[43m.\u001b[49m\u001b[43mCreateProcess\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexecutable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1555\u001b[39m \u001b[43m                             \u001b[49m\u001b[38;5;66;43;03m# no special security\u001b[39;49;00m\n\u001b[32m   1556\u001b[39m \u001b[43m                             \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m   1557\u001b[39m \u001b[43m                             \u001b[49m\u001b[38;5;28;43mint\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mclose_fds\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1558\u001b[39m \u001b[43m                             \u001b[49m\u001b[43mcreationflags\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1559\u001b[39m \u001b[43m                             \u001b[49m\u001b[43menv\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1560\u001b[39m \u001b[43m                             \u001b[49m\u001b[43mcwd\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1561\u001b[39m \u001b[43m                             \u001b[49m\u001b[43mstartupinfo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1562\u001b[39m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[32m   1563\u001b[39m     \u001b[38;5;66;03m# Child is launched. Close the parent's copy of those pipe\u001b[39;00m\n\u001b[32m   1564\u001b[39m     \u001b[38;5;66;03m# handles that only the child should have open.  You need\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m   1567\u001b[39m     \u001b[38;5;66;03m# pipe will not close when the child process exits and the\u001b[39;00m\n\u001b[32m   1568\u001b[39m     \u001b[38;5;66;03m# ReadFile will hang.\u001b[39;00m\n\u001b[32m   1569\u001b[39m     \u001b[38;5;28mself\u001b[39m._close_pipe_fds(p2cread, p2cwrite,\n\u001b[32m   1570\u001b[39m                          c2pread, c2pwrite,\n\u001b[32m   1571\u001b[39m                          errread, errwrite)\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [WinError 2] El sistema no puede encontrar el archivo especificado"
     ]
    }
   ],
   "source": [
    "# Crear la ruta del directorio de entrada y salida de datos\n",
    "output_dir = \"\"\n",
    "\n",
    "# Hacer el prefetch de los archivos srr\n",
    "subprocess.run([\"prefetch\", \"--output-directory\", output_dir, \"SRR031714\"], check=True)\n",
    "subprocess.run([\"prefetch\", \"--output-directory\", output_dir, \"SRR031715\"], check=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
